match/lib/match/portal_cache.rb

```diff
@@ -0,0 +1,114 @@
+          bundle_id_identifiers: bundle_id_identifiers,
+          needs_profiles_devices: ProfileIncludes.can_force_include?(params: params, notify: true) && !params[:force] && !params[:readonly],
+          needs_profiles_certificate_content: !ProfileIncludes.can_force_include_all_certificates?(params: params),
+          include_mac_in_profiles: params[:include_mac_in_profiles]
+        )
+
+        return cache
+      end
+
+      attr_reader :platform, :profile_type, :bundle_id_identifiers, :additional_cert_types, :needs_profiles_devices, :needs_profiles_certificate_content, :include_mac_in_profiles
+
+      def initialize(platform:, profile_type:, additional_cert_types:, bundle_id_identifiers:, needs_profiles_devices:, needs_profiles_certificate_content:, include_mac_in_profiles:)
+        @platform = platform
+        @profile_type = profile_type
+
+        # Bundle Ids
+        @bundle_id_identifiers = bundle_id_identifiers
+
+        # Certs
+        @additional_cert_types = additional_cert_types
+
+        # Profiles
+        @needs_profiles_devices = needs_profiles_devices
+        @needs_profiles_certificate_content = needs_profiles_certificate_content
+
+        # Devices
+        @include_mac_in_profiles = include_mac_in_profiles
+      end
+
+      def portal_profile(stored_profile_path:, keychain_path:)
+        parsed = FastlaneCore::ProvisioningProfile.parse(stored_profile_path, keychain_path)
+        uuid = parsed["UUID"]
+
+        portal_profile = self.profiles.detect { |i| i.uuid == uuid }
+
+        portal_profile
+      end
+
+      def reset_certificates
+        @certificates = nil
+      end
+
+      def forget_portal_profile(portal_profile)
+        return unless @profiles && portal_profile
+
+        @profiles -= [portal_profile]
+      end
+
+      def bundle_ids
+        return @bundle_ids.dup if @bundle_ids
+
+        @bundle_ids = Match::Portal::Fetcher.bundle_ids(
+          bundle_id_identifiers: @bundle_id_identifiers
+        )
+
+        return @bundle_ids.dup
+      end
+
+      def certificates
+        return @certificates.dup if @certificates
+
+        @certificates = Match::Portal::Fetcher.certificates(
+          platform: @platform,
+          profile_type: @profile_type,
+          additional_cert_types: @additional_cert_types
+        )
+
+        return @certificates.dup
+      end
+
+      def profiles
+        return @profiles.dup if @profiles
+
+        @profiles = Match::Portal::Fetcher.profiles(
+          profile_type: @profile_type,
+          needs_profiles_devices: @needs_profiles_devices,
+          needs_profiles_certificate_content: @needs_profiles_certificate_content
+        )
+
+        return @profiles.dup
+      end
+
+      def devices
+        return @devices.dup if @devices
+
+        @devices = Match::Portal::Fetcher.devices(
+          platform: @platform,
+          include_mac_in_profiles: @include_mac_in_profiles
+        )
+
+        return @devices.dup
+      end
```

Comment:
I think all of these blocks should be equivalent to the slightly more readable/idiomatic (example):

```
def devices 
  @devices ||= begin 
    Match::Portal::Fetcher.devices(
      platform: @platform,
      include_mac_in_profiles: @include_mac_in_profiles
    )
  end.dup
end
```

Reply:
Thanks for the suggestion. Done in 6a3d85be9454d36681930a794b40adaebdf0e860.

1. Done better assignment to a variable with `||=`.
2. Return @devices.dup statement stays intact.

The cache must be controlled in one class. Since the array is an object and open for mutations, I make a dupe on return.
Your solution copies only the fetch result; all external modifications impact the cached array.

Test code:
```ruby
def devices_a
  # This dupes only fetch, not devices
  @devices_a ||= begin
    puts "Fetch"
    ["a", "b"]
  end.dup
end
devices_a_1 = devices_a
puts "devices_a_1: #{devices_a_1}"
devices_a_1 << "c"
puts "devices_a_1: #{devices_a_1}"
devices_a_2 = devices_a
puts "devices_a_2: #{devices_a_2}"
puts "devices_a_2 is ['a', 'b'] #{devices_a_2 == ["a", "b"]}"

def devices_b
  # This dupes devices
  @devices_b ||= begin
    puts "Fetch"
    ["a", "b"]
  end

  return @devices_b.dup
end
devices_b_1 = devices_b
puts "devices_b_1: #{devices_b_1}"
devices_b_1 << "c"
puts "devices_b_1: #{devices_b_1}"
devices_b_2 = devices_b
puts "devices_b_2: #{devices_b_2}"
puts "devices_b_2 is ['a', 'b'] #{devices_b_2 == ["a", "b"]}"
```

Output:
```
Fetch
devices_a_1: ["a", "b"]
devices_a_1: ["a", "b", "c"]
devices_a_2: ["a", "b", "c"]
devices_a_2 is ['a', 'b'] false
Fetch
devices_b_1: ["a", "b"]
devices_b_1: ["a", "b", "c"]
devices_b_2: ["a", "b"]
devices_b_2 is ['a', 'b'] true
```